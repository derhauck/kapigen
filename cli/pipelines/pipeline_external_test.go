package pipelines_test

import (
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"gitlab.com/kateops/kapigen/cli/factory"
	"gitlab.com/kateops/kapigen/cli/internal/cli"
	"gitlab.com/kateops/kapigen/cli/internal/pipeline/config"
	"gitlab.com/kateops/kapigen/cli/internal/version"
	"gitlab.com/kateops/kapigen/cli/pipelines"
	"gitlab.com/kateops/kapigen/cli/types"
	"gitlab.com/kateops/kapigen/dsl/gitlab/job"
	"gitlab.com/kateops/kapigen/dsl/gitlab/stages"
)

type TestPipelineConfig struct {
	ConfigId string `yaml:"id"`
}

func (t *TestPipelineConfig) New() types.PipelineConfigInterface {
	return &TestPipelineConfig{}
}
func (t *TestPipelineConfig) Validate() error {
	return nil
}

func (t *TestPipelineConfig) Build(_ *factory.MainFactory, _ types.PipelineType, _ string) (*types.Jobs, error) {
	return &types.Jobs{
		types.NewJob("test", "alpine", func(ciJob *job.CiJob) {
			ciJob.TagMediumPressure().
				AddScript("echo 'hello world'").
				SetStage(stages.TEST)
		}),
	}, nil
}

func (t *TestPipelineConfig) Rules() *job.Rules {
	return job.DefaultMainBranchRules()
}

func TestExtendPipelines(t *testing.T) {
	t.Run("extend pipeline functions", func(t *testing.T) {
		const TestPipeline types.PipelineType = "test"

		pipelines.ExtendPipelines(map[types.PipelineType]types.PipelineConfigInterface{
			TestPipeline: &TestPipelineConfig{},
		})
		settings := cli.NewSettings(
			cli.SetMode(version.GetModeFromString(version.Gitlab.Name())),
		)

		pipelineConfig := &types.PipelineConfig{
			Pipelines: []types.PipelineTypeConfig{
				{
					Type: TestPipeline,
					Config: TestPipelineConfig{
						ConfigId: "configId",
					},
					PipelineId: "testId",
				},
			},
		}

		pipelineJobs, err := types.LoadJobsFromPipelineConfig(factory.New(settings), pipelineConfig, config.PipelineConfigTypes)
		if err != nil {
			t.Error(err)
		}
		if len(pipelineJobs.GetJobs()) == 0 {
			t.Error("no jobs found")
		}

		snaps.MatchSnapshot(t, pipelineJobs)

	})
}
