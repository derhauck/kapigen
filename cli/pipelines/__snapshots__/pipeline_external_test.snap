
[TestExtendPipelines/extend_pipeline_functions - 1]
&types.Jobs{
    &types.Job{
        Names: {"test", "testId", "test"},
        CiJob: &job.CiJob{
            Artifact:    job.Artifacts{},
            AfterScript: job.AfterScript{
                Value: &wrapper.Array[string]{
                    slice: {},
                },
            },
            BeforeScript: job.BeforeScript{
                Value: &wrapper.Array[string]{
                    slice: {},
                },
            },
            Script: job.Script{
                Value: &wrapper.Array[string]{
                    slice: {"echo 'hello world'"},
                },
            },
            AllowFailure: job.AllowFailure{},
            Cache:        job.Cache{
                Key:   "",
                Paths: wrapper.Array[string]{
                    slice: {},
                },
                Unprotect: true,
                Policy:    2,
                Active:    false,
            },
            Variables: {},
            Tags:      {
                &job.Tag{},
            },
            Image: job.Image{
                Name:       "alpine",
                Entrypoint: wrapper.Array[string]{},
                PullPolicy: 0,
            },
            Rules: {
                &job.Rule{
                    If:           "$KTC_STOP_PIPELINE != \"false\" && $DEBUG == null",
                    Changes:      wrapper.Array[string]{},
                    AllowFailure: false,
                    Variables:    struct {}{},
                    When:         job.WhenWrapper{
                        Value: &enum.When(3),
                    },
                },
                &job.Rule{
                    If:           "$CI_DEFAULT_BRANCH == $CI_COMMIT_BRANCH",
                    Changes:      wrapper.Array[string]{},
                    AllowFailure: false,
                    Variables:    struct {}{},
                    When:         job.WhenWrapper{
                        Value: &enum.When(0),
                    },
                },
                &job.Rule{
                    If:           "$KTC_TEST_PIPELINE",
                    Changes:      wrapper.Array[string]{},
                    AllowFailure: false,
                    Variables:    struct {}{},
                    When:         job.WhenWrapper{},
                },
            },
            Stage:    3,
            Services: job.Services{
                Values: &[]*job.Service{
                },
            },
            Coverage: "",
        },
        Needs: {
        },
        currentName: 2,
        CiJobYaml:   &job.CiJobYaml{
            Artifacts:    (*job.ArtifactsYaml)(nil),
            AfterScript:  {},
            AllowFailure: bool(false),
            BeforeScript: {},
            Cache:        (*job.CacheYaml)(nil),
            Script:       {"echo 'hello world'"},
            Needs:        &job.NeedsYaml(nil),
            Variables:    {},
            Image:        &job.ImageYaml{
                Name:       "alpine",
                Entrypoint: nil,
                PullPolicy: "always",
            },
            Rules: &job.RulesYaml{
                &job.RuleYaml{
                    If:           "$KTC_STOP_PIPELINE != \"false\" && $DEBUG == null",
                    Changes:      nil,
                    AllowFailure: bool(false),
                    Variables:    struct {}{},
                    When:         "never",
                },
                &job.RuleYaml{
                    If:           "$CI_DEFAULT_BRANCH == $CI_COMMIT_BRANCH",
                    Changes:      nil,
                    AllowFailure: bool(false),
                    Variables:    struct {}{},
                    When:         "on_success",
                },
                &job.RuleYaml{
                    If:           "$KTC_TEST_PIPELINE",
                    Changes:      nil,
                    AllowFailure: bool(false),
                    Variables:    struct {}{},
                    When:         "on_success",
                },
            },
            Stage:    "dynamic",
            Services: &job.ServiceYamls{
            },
            Tags:     {"pressure:medium"},
            Coverage: "",
        },
        PipelineId:   "testId",
        ExternalTags: nil,
    },
}
---
